cpp/lib/ome/bioformats/MetadataTools.cpp

```diff
@@ -494,6 +566,142 @@ namespace ome
+      if (root)
+        {
+          std::shared_ptr<ome::xml::model::Image>& imageref(root->getImage(image));
+          if (image)
+            {
+              std::shared_ptr<ome::xml::model::Pixels> pixels(imageref->getPixels());
+              if (pixels)
+                {
+                  std::vector<std::shared_ptr<ome::xml::model::Channel> > channels(pixels->getChannelList());
+                  for (Metadata::index_type c = 0U; c < channels.size(); ++c)
+                    {
+                      std::shared_ptr<ome::xml::model::Channel> channel(channels.at(c));
+                      if (channel->getID().empty() || c >= sizeC)
+                        pixels->removeChannel(channel);
+                    }
+                }
+            }
+        }
+    }
+
+    MetadataMap
+    getOriginalMetadata(::ome::xml::meta::OMEXMLMetadata& omexml)
+    {
+      MetadataMap map;
+
+      std::shared_ptr<ome::xml::meta::OMEXMLMetadataRoot> root(std::dynamic_pointer_cast<ome::xml::meta::OMEXMLMetadataRoot>(omexml.getRoot()));
+      if (root)
+        {
+          std::shared_ptr<StructuredAnnotations> sa(root->getStructuredAnnotations());
+          if (sa)
+            {
+              for (OMEXMLMetadata::index_type i = 0; i < sa->sizeOfXMLAnnotationList(); ++i)
+                {
+                  // Check if this is an OriginalMetadataAnnotation object.
+                  std::shared_ptr<XMLAnnotation> annotation(sa->getXMLAnnotation(i));
+                  std::shared_ptr<OriginalMetadataAnnotation> original(std::dynamic_pointer_cast<OriginalMetadataAnnotation>(annotation));
+                  if (original)
+                    {
+                      const OriginalMetadataAnnotation::metadata_type kv(original->getMetadata());
+                      map.set(kv.first, kv.second);
+                      continue;
+                    }
+
+                  // Fall back to parsing by hand.
+                  try
+                    {
+                      xerces::Platform xmlplat;
+                      xerces::dom::ParseParameters params;
+                      params.validationScheme = xercesc::XercesDOMParser::Val_Never;
+                      xerces::dom::Document doc(ome::xerces::dom::createDocument(annotation->getValue()));
+
+                      std::vector<xerces::dom::Element> OriginalMetadataValue_nodeList = ome::xml::model::detail::OMEModelObject::getChildrenByTagName(doc, "OriginalMetadata");
+                      if (OriginalMetadataValue_nodeList.size() > 1)
+                        {
+                          format fmt("Value node list size %1% != 1");
+                          fmt % OriginalMetadataValue_nodeList.size();
+                          throw ModelException(fmt.str());
+                        }
+                      else if (OriginalMetadataValue_nodeList.size() != 0)
```

Comment:
Yes, and I'd like to change it at some point.  However, it's a copy of the logic in the xsd-fu templates-java/OMEXMLModelObject.template for a custom derived model object, so I'd rather change it across the board rather than making it inconsistent in one place.  Happy to do that here if @melissalinkert is OK with updating the xsd-fu templates as well.